### ISSUE TITLE
Hiera as a Service (HAAS) proposal

### ISSUE THREAD

**Issue by timidri** [lyraproj/hiera#46]
# Summary 

Many teams who use Puppet to manage server-based infrastructure are also starting to manage cloud-native application delivery including the needed infrastructure. Usually, those teams have codified configuration data in hiera for the Puppet node-based configuration processes, but they cannot use this company-wide configuration data in a cloud-native deployment pipeline. If Hiera would be available as a service, hiera lookups could be easily done outside the "traditional Puppet context", for example from a delivery pipeline, and enable teams to define configuration data only once - inside hiera.

# Proposed approach

1. Finalize clean room implementation of hiera including pluggable backend system and a yaml backend
1. Develop a network front-end to hiera (essentially introducing _hiera-server_)
1. Develop a pluggable RBAC system to the _hiera server_

_Note:  HAAS means "hare" in Dutch_

**Comment by thallgren** [comment]
I think that #1 and #2 are already in place. Hiera can be used as a service. There's also a docker image for the hiera micro service: https://hub.docker.com/r/lyraproj/hiera.

I'm not sure if RBAC should be an integral part of the hiera micro service. Would it be possible to use RBAC to control access to the docker image?

**Comment by timidri** [comment]
@thallgren Great, I had this lying around for a while so I didn't notice the microservice has already been implemented. I just tried it out and the yaml lookups work great!

It seems to me RBAC service would be a requirement if the hiera microservice would run on the Puppet master providing access to the company-wide hiera config and access control to the docker image would be irrelevant.

**Comment by thallgren** [comment]
I didn't mean access to the image per se, but to a running instance of it :-) I.e. could Hiera run on a separate docker instance and have RBAC control access to that instance? Or would it be a requirement that the actual hiera process would run directly on the Puppet master?

**Comment by timidri** [comment]
Well, that will depend on the hiera backend. The majority of hiera users use the yaml backend which means that the hiera server would need access to the filesystem.

**Comment by thallgren** [comment]
Ok. I guess that I could make Hiera aware that an optional RBAC service might be present.

Some questions:
Should different roles have access to different hierarchies, or even keys, in hiera? Or is it more of a "access or no access to the hiera service" kind of thing?
Do you have any examples of which RBAC solutions that would be desirable to plug in?

**Comment by timidri** [comment]
1. MVP would be to have "access or no access" to the hiera service.
2. As a first enhancement, it would be great to be able to use a valid PE RBAC token to get access to the hiera service
3. [RBAC solution used by Hashicorp Vault](https://www.vaultproject.io/docs/concepts/policies.html) seems to be a good candidate or at least a sensible model to integrate

**Comment by ahpook** [comment]
I'd strongly prefer to have some kind of front end manage the access control, rather than hiera itself. It could use this codebase as a library to do the actual lookups, but manage access on its own.

Put another way, an RBAC implementation is way out of scope for this repo's design goals.

**Comment by ahpook** [comment]
Going to close this issue since the basic question of a REST-accessible API for Hiera is addressed.
